tutorials/e-commerce-store-codelab/step-07-checking-out.md

```diff
@@ -0,0 +1,167 @@
+---
+description: Implementing checking out and order placement for an e-commerce app built with Remix, Drizzle, Turso, and Cloudflare Workers
+keywords:
+  - turso
+  - remix
+  - e-commerce
+---
+
+# Step 7: Checking out and placing orders
+
+After users have selected items they want to buy and placed them in the cart, we
+then expect them to checkout. To facilitate this inside the app, let’s create
+the checkout page by adding a `checkout._index.tsx` file under `app/routes`.
+
+For the loader, we are trying to get the same data as we did in the cart page, only that here we need the user information too. The user data is used to pre-fill some fields on the checkout form.
+
+```tsx
+cartItems = await db.query.cartItems.findMany({
+  where: (cartItems, { eq }) => eq(cartItems.userId, userId),
+  columns: { count: true, id: true },
+  with: { product: true, user: true },
+});
+return {
+  cartItems,
+  userData: cartItems[0].user,
+};
+```
+
+We then get the loader data and lay out the cart items and checkout information
+on the page component. [Click here] to see the page markup of the checkout page
+component.
+
+After having registered a user and added some items to the cart, this is what
+you should see when you visit the checkout page.
+
+![Checkout page preview](assets/10-checkout-page-preview.png)
+
+After having provided the required payment, contact, and billing information,
+the user will submit the checkout form by clicking the “Pay Now” button.
+
+Clinking the "Pay Now" button prompts the submission
+of the checkout form data to the checkout page's action that validates submitted data, handles the creation
+of new orders, and returns the correct message depending on the action being
+successful or not.
+
+Here is the page action that handles order placement.
+
+```tsx
+// app/routes/checkout._index.tsx
+
+export const action = async ({
+  request,
+  context,
+}: ActionArgs): Promise<
+  { status: string; message: string; data: any } | Response
+> => {
+  const userId = await requireUserId(
+    { request, redirectTo: "/account/login" },
+    context
+  );
+  if (userId === undefined) {
+    return redirect("/account/login");
+  } else {
+    const formData = await request.formData();
+    const values = Object.fromEntries(formData);
+    // look for missing input
+    if (
+      !values.firstName ||
+      !values.lastName ||
+      !values.email ||
+      !values.phone ||
+      !values.country ||
+      !values.zipCode
```

Comment:
A string value of nothing but whitespace for any one of these would probably be considered missing data, but not evaluate false. In order to avoid misleading people, either intentionally skip validation checks with a comment, or find a way to be more thorough.  Properly thorough validation of each of these types of data requires extreme detail, though.